pandas/_libs/algos_rank_helper.pxi.in

```diff
@@ -123,17 +127,23 @@ def rank_1d_{{dtype}}(object in_arr, ties_method='average', ascending=True,
         _as = _as[::-1]
 
     sorted_data = values.take(_as)
+    # need to distinguish between pos/neg nan and real nan when keep_na is true
+    {{if dtype != 'uint64'}}
+    sorted_namask = mask.take(_as)
+    sorted_namask = sorted_namask.astype(np.bool)
```

Comment:
this is a very odd thing to do, are you just computing the mask?

Reply:
No, I am sorting the mask to match the sorted_data. When I loop through the sorted_data when computing ranks, I need to find the corresponding indicator for nan value. The current version is comparing the each element against nan_value, which is causing the problem
```
if (val is nan_value) and keep_na:
    ranks[argsorted[i]] = nan
```

Because nan_value is either pos or neg infinite value.
```
    if ascending ^ (na_option == 'top'):
        nan_value = {{pos_nan_value}}
    else:
        nan_value = {{neg_nan_value}}
```
As a result when the underlying value is truely +/- infinite, the rank will be just nan for both. That's why ranking with inf values are failing when keep_na is true. Can you elaborate more on why this is odd ? Do you mean the `sorted_namask.astype(np.bool)` is redundant ?